ESLint's cache can stop autofixing

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** 5.2.0
* **Node Version:** v10.7.0
* **npm Version:** 6.1.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default or babel-eslint
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```json
{
  "rules": {
    "quotes": 1
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
'test'
```

<!-- Paste the command you used to run ESLint: -->
```bash
.\node_modules\.bin\eslint --cache src
.\node_modules\.bin\eslint --fix --cache src
```

**What did you expect to happen?**
ESLint should fix the fixable issues.

**What actually happened? Please include the actual, raw output from ESLint.**
```
C:\test> .\node_modules\.bin\eslint --cache src

C:\test\src\test.js
  1:1  warning  Strings must use doublequote  quotes

✖ 1 problem (0 errors, 1 warning)
  0 errors and 1 warning potentially fixable with the `--fix` option.

C:\test> .\node_modules\.bin\eslint --fix --cache src

C:\test\src\test.js
  1:1  warning  Strings must use doublequote  quotes

✖ 1 problem (0 errors, 1 warning)
  0 errors and 1 warning potentially fixable with the `--fix` option.
```
ESLint does not autofix when the cache is populated. If I delete `.eslintcache`, then ESLint will autofix just fine.
I believe this is a regression since 5.2.0 (81283d0).
ESLint, when autofixing, shouldn't ignore files that are cached, do not invalidate the cache, but are known to have issues.

--- a/lib/cli-engine.js
+++ b/lib/cli-engine.js
@@ -518,18 +518,18 @@ class CLIEngine {
             }
 
             if (options.cache) {
-
-                /*
-                 * get the descriptor for this file
-                 * with the metadata and the flag that determines if
-                 * the file has changed
-                 */
                 const cachedLintResults = lintResultCache.getCachedLintResults(fileInfo.filename);
 
                 if (cachedLintResults) {
-                    debug(`Skipping file since it hasn't changed: ${fileInfo.filename}`);
+                    const resultHadMessages = cachedLintResults.messages && cachedLintResults.messages.length;
 
-                    return cachedLintResults;
+                    if (resultHadMessages && options.fix) {
+                        debug(`Reprocessing cached file to allow autofix: ${fileInfo.filename}`);
+                    } else {
+                        debug(`Skipping file since it hasn't changed: ${fileInfo.filename}`);
+
+                        return cachedLintResults;
+                    }
                 }
             }
 

--- a/tests/lib/cli-engine.js
+++ b/tests/lib/cli-engine.js
@@ -1473,6 +1473,31 @@ describe("CLIEngine", () => {
                 });
             });
 
+            it("should run autofix even if files are cached without autofix results", () => {
+                const baseOptions = {
+                    cwd: path.join(fixtureDir, ".."),
+                    useEslintrc: false,
+                    rules: {
+                        semi: 2,
+                        quotes: [2, "double"],
+                        eqeqeq: 2,
+                        "no-undef": 2,
+                        "space-infix-ops": 2
+                    }
+                };
+
+                engine = new CLIEngine(Object.assign({}, baseOptions, { cache: true, fix: false }));
+
+                // Do initial lint run and populate the cache file
+                engine.executeOnFiles([path.resolve(fixtureDir, `${fixtureDir}/fixmode`)]);
+
+                engine = new CLIEngine(Object.assign({}, baseOptions, { cache: true, fix: true }));
+
+                const report = engine.executeOnFiles([path.resolve(fixtureDir, `${fixtureDir}/fixmode`)]);
+
+                assert.ok(report.results.some(result => result.output));
+            });
+
         });
 
         // These tests have to do with https://github.com/eslint/eslint/issues/963

